test/built-ins/Promise/race/resolved-sequence-extra-ticks.js

```diff
@@ -0,0 +1,41 @@
+// Copyright (C) 2020 Rick Waldron. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-promise.race
+description: Resolution ticks are set in a predictable sequence with extra then calls
+info: |
+  PerformPromiseRace
+
+  Repeat,
+    Let next be IteratorStep(iteratorRecord).
+    If next is an abrupt completion, set iteratorRecord.[[Done]] to true.
+    ReturnIfAbrupt(next).
+    If next is false, then
+      Set iteratorRecord.[[Done]] to true.
+      Return resultCapability.[[Promise]].
+    Let nextValue be IteratorValue(next).
+    If nextValue is an abrupt completion, set iteratorRecord.[[Done]] to true.
+    ReturnIfAbrupt(nextValue).
+    Let nextPromise be ? Call(promiseResolve, constructor, « nextValue »).
+    Perform ? Invoke(nextPromise, "then", « resultCapability.[[Resolve]], resultCapability.[[Reject]] »).
+
+flags: [async]
+includes: [compareArray.js,promiseHelper.js]
+---*/
+
+let a = new Promise(resolve => resolve({}));
+let sequence = [1];
+Promise.all([
+  Promise.race([a]).then(resolved => {
+    sequence.push(4);
+  }),
+  a.then(() => {
+    sequence.push(3);
+  }).then(() => {
+    sequence.push(5);
+  }),
+]).then(() => {
+  checkSequence(sequence);
```

Comment:
Here and other tests: checkSequence only checks that "sequence" is [1], [1, 2], [1, 2, 3] etc, but it doesn't assert that we get a full sequence. If for some reason, we we expecting [1, 2, 3, 4, 5] but only get [1, 2], the test would pass. Would it make sense to add a sequence length parameter to checkSequence?

Reply:
To prove my point, this kind of testing error happens in test/built-ins/Promise/race/resolved-sequence-mixed.js . It only calls checkSequence([1, 2, 3]) and checkSequence([1, 2, 3, 4]), even though from the test it looks like it would build a longer sequence.

Also we should call the final checkSequence in a place which is surely executed. Looks like that test also suffers from the problem that the checkSequence is inside dead code.

In addition, we should add a reject handler which asserts into all places which don't expect a rejection. Otherwise the rejections are just silently swallowed.

Reply:
When I add `throw 1` before line 39, this test will fail, so there is no dead code there. The presence of `.then($DONE, $DONE);` ensures that no rejections get swallowed. Anything that passes through `$DONE`, which is not `undefined`, is treated as a failure message.

Reply:
> The presence of .then($DONE, $DONE); ensures that no rejections get swallowed. Anything that passes through $DONE, which is not undefined, is treated as a failure message.

You can test this by putting `throw 1;` anywhere in the test: 

```js
let a = new Promise(resolve => {throw 1; resolve({})});
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    throw 1;
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
    throw 1;
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    throw 1;
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
    throw 1;
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    throw 1;
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
    throw 1;
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
    throw 1;
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
    throw 1;
  checkSequence(sequence);
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```

```js
let a = new Promise(resolve => resolve({}));
let sequence = [1];
Promise.all([
  Promise.race([a]).then(resolved => {
    sequence.push(4);
  }),
  a.then(() => {
    sequence.push(3);
  }).then(() => {
    sequence.push(5);
  }),
]).then(() => {
  assert.sameValue(sequence.length, 5);
  checkSequence(sequence);
    throw 1;
}).then($DONE, $DONE);
sequence.push(2);
```

Result : 
```
FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (default)
  1

FAIL test/built-ins/Promise/race/resolved-sequence-extra-ticks.js (strict mode)
  1
```